### ISSUE TITLE
Write documentation for roles & permissions concept

### ISSUE THREAD

**Issue by kulmann** [owncloud/ocis-settings#13]
Write documentation for roles & permissions concept

**Comment by refs** [comment]
Preliminary draft:

## Roles and Permissions

Roles and Permissions live in the `ocis-settings` service. There is a set of basic operations that are defined on the settings service regarding roles and permissions, and these are:

- ListRoles
- ListRoleAssignments
- AssignRoleToUser
- RemoveRoleFromUser

The scope of roles and permissions is only inside the `ocis-extensions` framework. Do not confuse with [Reva grants](https://cs3org.github.io/cs3apis/#cs3.sharing.link.v1beta1.Grant). Roles and permissions do not handle actions on shared resources. i.e: an admin cannot restrict access to users on a given folder based on roles defined in this service.

The intended usage of Roles and Permissions can be clarified with an example: An admin user wants to configure so that only users with the role `USER` can update their name. This means that only users with roles `ADMIN` and `USER` (since a user with the `ADMIN` role contains the role `USER`) will be allowed to update such setting on the `ocis-accounts` extension. In this scenario, users with the role `GUEST` will not be allowed to update such option.

## Assigning a role to a user

When assigning a role to a user an **assignment** is created. An assignment relates an account with a role.

## Design

Roles and Permissions was built on top of settings bundles. This API reuses most of what has already been implemented for the settings bundles and extends it with roles and permissions. It does so by adding the `TYPE_ROLE` to the definition of a `SettingsBundle` and also a value of type `PermissionSetting` to the `Setting` message type.

When an extension starts, it registers its roles and permissions (as well as the settings bundles, in fact, roles and permissions are just another settings bundle of type `ROLE`) to the `ocis-settings` service (rendering `ocis-settings` a mandatory service to start on the runtime) that records which roles are allowed to change the extension behavior.

### TODO:

- insert diagrams: https://cloud.owncloud.com/index.php/s/qcq08xjN9WuSZKS

## Vocabulary

**Role**: A named set of permissions.

**Permission**: An operation that can be done on a constraint. i.e: "update my language on the files extension".

**Assignment**: Tuple of account_uuid + role_id. Internal.

## TODO

- validation is needed when persisting a type role on a settings bundle.

**Comment by kulmann** [comment]
Changes:
- added a note on the exposed uuids of the default roles
- split the micro calls into `request` and `response`
- added details about `bundles` (below the protobuf message)
- added a short paragraph on `values` (at the bottom)

**Comment by refs** [comment]
User stories:
- As an admin I want to be able to assign and manage roles to users.
- As an admin I want to be able to define roles on a per extension base:
  - i.e: I want to create a new role only valid for the `XXXXXX` extension.

Open questions:
- If we decide to create a superset of roles on top of Reva grants we need to clearly define how are we going to handle them:
  - do `ocis-settings` roles take precedence?
    - this would mean if user A shares a folder with user B with `RW` permission but we have a role on the files app that states user B has only `R` permissions, which one takes precedence?
  - mapping Reva grants translates into glue code and defining which combination of [grant permissions](https://cs3org.github.io/cs3apis/#cs3.storage.provider.v1beta1.ResourcePermissions) constitute a `R` only on `ocis-settings`.

**Comment by refs** [comment]
@kulmann ☝️

**Comment by refs** [comment]
- Can any extension create as many roles as it wants?
- We need to validate extension name
- When fetching values:
  - user value present, system value not: return user value
  - user value not present, system value present: return system value
  - user value present, system value present: return user value (user takes precedence)

**Comment by butonic** [comment]
@refs @kulmann reading the draft i stumbled over
> ... (since a user with the ADMIN role contains the role USER) ...

This implies either role inheritance or that a user can have multiple roles. Can you clarify the text?

**Comment by refs** [comment]
@butonic that's correct. I thought of role inheritance as well, which with this model is a bit hard to map but it was something that was mentioned and I remembered it, therefore added to the doc. It is another point that we'd need to discuss for the MVP.

**Comment by kulmann** [comment]
For the mvp I‘d advise against role inheritance. A user can have as many roles as you want - data model wise. As long as we’re still learning about requirements it makes sense to compose each of the default roles with their respective full set of permissions. For the UI we agreed to only allow single role assignment.

Having multiple roles only makes sense then when we‘re coming up with e.g. group roles. Which is technically possible already.
